### ISSUE TITLE
Handle large data sets

### ISSUE THREAD

**Issue by marktheunissen** [vole/vole#20]
Some ideas:

As the number of posts grows, the number of files will become unmanageable. There are ways to mitigate this, for example, we could do what git & others do and store things in directories named after the first two letters of the uuid.

```
/ab/1234567890-post-abdsf-sdfwf-wefwef-wefwef
/ab/1234567890-post-abvds-dfvb-dfghtryt-rtyryrt
/qd/1234567890-post-qddsf-sdfwf-wefwef-wefwef
```

This is handy but makes it harder to find the most recent posts ordered by timestamp. Perhaps we have the most recent stuff in it's own folder, then do housekeeping to move older content into a folder structure like this.

We may also want to introduce a concept of archived content, perhaps configurable by the user. I may choose to discard anything older than 5 months, or cap the maximum directory size at a limit.

We could also compress older content automatically. Then, when other users receive a .tar.gz file as part of someone's vole data directory, their Vole will extract this archive and show the posts, if requested to. This cuts down on bandwidth.

Important to watch out for malware that may try exploit a user's Vole this way. We should be very strict about parsing input.

**Comment by marktheunissen** [comment]
Other ideas talked about:
1. Use a sqlite database for an index. Possible problems with this over bittorrent. We could experiment but I suspect any database file like this may become corrupted (on the read-only side). Also, the bittorrent protocol will completely re-send any changed file under 4M (Bittorrent FAQ has this I believe.)

**Comment by pifantastic** [comment]
I guess I was thinking we wouldn't sync the database file at all. It would just be used as an index for incoming files. Although that approach might be more trouble than it's worth.

**Comment by marktheunissen** [comment]
Oh I see, almost like a cache of the data. Definitely adds complexity and not sure if the payoffs are big enough.

**Comment by marktheunissen** [comment]
I've been thinking more about data storage. I guess using a 'real' embedded database would be ok when there is only one writer to the database - as the 'followers' will be read-only, and thus there is no chance of two copies of the data file being written to at the same time.

But if there are two users with write access to an account, one could easily destroy the other's work, because bittorrent sync protocol only saves the most recent copy of any file. 

And in a 'groups' situation, everyone with read/write to the folder can write entries, thus plenty of opportunity for data loss if it's a single file everyone is writing to. This is why the uuid system works so well in this context.

We almost need something like SQLite, LevelDB, Kyoto Cabinet, that writes to individual files

We could also build in some eventually-consistent system but that requires knowing at least how many clients there are with write access to a datastore, surely?

**Comment by marktheunissen** [comment]
I would describe what we need as a "fault-tolerant datastore or hash table, backed by an untrustable filesystem that handles the eventual consistency across all clients".

**Comment by marktheunissen** [comment]
How about sharding by timestamp and by uuid?

Every 4 million seconds make a new directory.

```
/posts/10000000000/ec/1368990899386735016-post-ec41d512-1c6a-4fa7-5c52-b3920b7f9acf
/posts/11000000000/da/1369015076104026988-post-dabb2ca6-c88b-4404-5f69-b70b9c6a6b3f
...
```

That makes it easy to find the newest content and prevents a single directory from exploding.

**Comment by pifantastic** [comment]
The problem with timestamps is that they're less likely to be evenly distributed, right? Using the first 2 chars of a SHA has is a good bet since the whole purpose of the hash is to be evenly distributed. 

However, I can't think of a good alternative as our file access is inherently time based. I suppose the alternative is to keep track of how many posts have been written to a folder, and then create a new one whenever it hits what we consider to be the max. But that loses the elegance of a method like what git does.

**Comment by Fil** [comment]
if the user is going to use their contents folder directly (think grep), then a reasonable sharding would be by YYYY/YYYY-MM/YYYY-MM-DD/ or just YYYY/YYYY-MM/

**Comment by Fil** [comment]
Also, please don't enforce this — accept ANY type of sharding, so that each user / client can experiment with their own ideas

**Comment by marktheunissen** [comment]
@Fil The sharding will have to be done in a way that minimizes the amount of disk reads that are required to load the most recent posts. I don't think we can just support any sharding as the structure of the data will carry meaning to the application.

We do plan on versioning Vole data, so right now you'll see a 'v1' folder which means the data is structured in a set way, once this changes we'll migrate to a v2 folder so older clients know that the data has changed and can prompt for an upgrade.

**Comment by Fil** [comment]
we could set a rule saying "the last folder in alphanum sort" will be watched for the most recent posts
but I thought you'd use something like inotify

**Comment by marktheunissen** [comment]
Yeah, that would work fine. Using inotify is planned, but on first startup it's necessary to scan the disk for posts in some way. If we accept any kind of sharding, as you have suggested, then someone could shard each post into a random folder, meaning that we'd have to scan every single directory and file.
